### DISCUSSION TITLE
How can I differentiate a query based on RLS from a person who dont have access, and a person who just dont have any data

### DISCUSSION THREAD

**Discussion by vongohren** [supabase/supabase#15671] [Questions] [answered]
I dont know how to differentiate a query, where a person does a SQL where RLS is enabled, if that person dont have access, or dont just have any data.

F.ex, there are two links
app/uuid1
app/uuid2

Person 1 has access to both and it loads two items on uuid2.
Person 2 gets ahold of uuid2 link, and loads it under his auth, but it does not load any data because of RLS

There is a difference here, and fair it can be differentiated.

But lets say uuid1 dont hold any data, so
Person 1 has access to uuid1 but it loads zero items on uuid1.
Person 2 gets ahold of uuid1 link, and loads it under his auth, but it does not load any data because of RLS.

The question is, how can I differentiate these two? For person 1 I want to show add data, but for person 2 i want to redirect or something. 

But the query result does not provide any indicators that the RLS policy has been invoked.

Can I do this in any way? Or am I thinking all wrong about this?

**Comment by GaryAustin1** [comment, marked as answer]
For security RLS on selects does not reveal any info about the request.  As far as the query is concerned the data looks like it does not exist, the same as if it does not.  You can't differentiate.

You might be able to do something in an rpc call and your own function code.

**Reply by vongohren** [reply]
@GaryAustin1 thanks for coming back to me, but then [the documentation](https://supabase.com/docs/learn/auth-deep-dive/auth-row-level-security) has to be updated, it is expected to return error, unless I read wrong?

But so what you are basiaclly saying, is that there is no way unless i do RPC and function code? Do you have any documentation on this?

**Reply by GaryAustin1** [reply]
https://www.postgresql.org/docs/current/sql-createpolicy.html  If you look at descriptions there it says insert returns an error which is true and then for example select does not return rows with no mention of error. Update will do nothing if USING does not pass but will error if WITH CHECK does not pass. So reading is silent but changing data, like insert will error. 

The one line I think you are referring to in Supabase doc is incorrect about reading returning an error. 

There are also many, many hits on the web and here in discussions.  Also see https://github.com/orgs/supabase/discussions/12895 .

**Reply by GaryAustin1** [reply]
The doc you referred to has been updated...
![image](https://github.com/supabase/supabase/assets/54564956/40345027-2d89-4b09-bec3-da54afae4206)

**Reply by vongohren** [reply]
Thanks for the fix and reply! Will se how I circumvent this :D
